scripts/update-message-registry.py

```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env python3
+
+from __future__ import annotations
+
+import argparse
+from dataclasses import astuple, dataclass
+from pathlib import Path
+from typing import Sequence
+
+import jinja2
+import yaml
+
+
+@dataclass(order=True)
+class Version:
+    """A semantic version number: MAJOR.MINOR.PATCH."""
+
+    UNKNOWN_VERSION = "UNKNOWN"
+    DEFAULT_VERSION = "1.0"  # [TODO] some initial version?
+
+    major: int
+    minor: int
+    patch: int = 0
+
+    def __str__(self):
+        return ".".join(map(str, astuple(self)))
+
+    @classmethod
+    def parse(cls, raw_version: str) -> Version:
+        if raw_version == cls.UNKNOWN_VERSION:
+            raw_version = cls.DEFAULT_VERSION
+        version = raw_version.split(".")
+        assert 2 <= len(version) <= 3 and all(
+            n.isdecimal() for n in version
+        ), raw_version
+        return Version(*map(int, version))
+
+
+@dataclass
+class Example:
+    """An example in a message entry."""
+
+    name: str
+    description: str
+    before: str | None
+    after: str | None
+
+    @classmethod
+    def parse(cls, entry) -> Example:
+        name = entry.get("name")
+        assert name, entry
+        description = entry.get("description")
+        assert description
+        before = entry.get("before")
+        after = entry.get("after")
+        # Either none or both of them
+        assert not (bool(before) ^ bool(after))
+        return Example(name=name, description=description, before=before, after=after)
+
+
+@dataclass
+class Entry:
+    """An xkbcommon message entry in the message registry"""
+
+    VALID_TYPES = ("warning", "error")
+
+    code: int
+    """A unique strictly positive integer identifier"""
+    id: str
+    """A unique short human-readable string identifier"""
+    type: str
+    """Log level of the message"""
+    description: str
+    """A short description of the meaning of the message"""
+    details: str
+    """A long description of the meaning of the message"""
+    added: Version
+    """Version of xkbcommon the message has been added"""
+    removed: Version | None
+    """Version of xkbcommon the message has been removed"""
+    examples: tuple[Example, ...]
+    """
+    Optional examples of situations in which the message occurs.
+    If the message is an error or a warning, also provide hints on how to fix it.
+    """
+
+    @classmethod
+    def parse(cls, entry) -> Entry:
+        code = entry.get("code")
+        assert code is not None and isinstance(code, int) and code > 0, entry
+        id = entry.get("id")
+        assert id is not None, entry
+        type_ = entry.get("type")
+        assert type_ in cls.VALID_TYPES, entry
+        description = entry.get("description")
+        assert description is not None, entry
+        details = entry.get("details", "")
+        raw_added = entry.get("added", "")
+        assert raw_added, entry
+        added = Version.parse(raw_added)
+        assert added, entry
```

Comment:
I think the above code would probably be a bit easier to read with some grouping of the `get` and the `asserts`
```python
code = entry.get("code")
id = entry.get("id")
....
assert code is not None and ...
assert entry is not None
....
```

same for `Example.parse` but there it's only two fieldsd

Reply:
I find it better when code is grouped semantically and when it fails early, to avoid unnecessary work. I added spaces to improve the reading. Is it good enough so?